src/System.Drawing.Common/src/System/Drawing/GdiPlusStreamHelper.Unix.cs

```diff
@@ -34,171 +34,94 @@
 // WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 //
 
+#if netcoreapp20
+using System.Buffers;
+#endif
+
 using System.IO;
 using System.Runtime.InteropServices;
 
 namespace System.Drawing
 {
     internal sealed partial class GdiPlusStreamHelper
     {
-        public Stream stream;
-
-        private StreamGetHeaderDelegate sghd = null;
-        private StreamGetBytesDelegate sgbd = null;
-        private StreamSeekDelegate skd = null;
-        private StreamPutBytesDelegate spbd = null;
-        private StreamCloseDelegate scd = null;
-        private StreamSizeDelegate ssd = null;
-        private byte[] start_buf;
-        private int start_buf_pos;
-        private int start_buf_len;
-        private byte[] managedBuf;
-        private const int default_bufsize = 4096;
+#if netcoreapp20
+        private const int DefaultBufferSize = 4096;
+#endif
+        private Stream _stream;
 
-        public GdiPlusStreamHelper(Stream s, bool seekToOrigin)
+        public unsafe GdiPlusStreamHelper(Stream stream, bool seekToOrigin)
         {
-            managedBuf = new byte[default_bufsize];
-
-            stream = s;
-            if (stream != null && stream.CanSeek && seekToOrigin)
+            // Seeking required
+            if (!stream.CanSeek)
             {
-                stream.Seek(0, SeekOrigin.Begin);
+                var memoryStream = new MemoryStream();
+                stream.CopyTo(memoryStream);
+                _stream = memoryStream;
             }
-        }
-
-        public int StreamGetHeaderImpl(IntPtr buf, int bufsz)
-        {
-            int bytesRead;
-
-            start_buf = new byte[bufsz];
-
-            try
+            else
             {
-                bytesRead = stream.Read(start_buf, 0, bufsz);
-            }
-            catch (IOException)
-            {
-                return -1;
+                _stream = stream;
             }
 
-            if (bytesRead > 0 && buf != IntPtr.Zero)
+            if (seekToOrigin)
             {
-                Marshal.Copy(start_buf, 0, (IntPtr)(buf.ToInt64()), bytesRead);
+                _stream.Seek(0, SeekOrigin.Begin);
```

Comment:
This seems wrong to me.  Maybe it doesn't matter based on usage patterns, but let's say stream.CanSeek is false and seekToOrigin is false.  Previously, the incoming stream would have had its position left as is, and subsequent calls to Stream.Read and the like would have processed from where it was.  But with this change, if CanSeek is false, the data from the original stream is going to be copied to the new MemoryStream, whose Position will then end up being after all of the data... with seekToOrigin also being false then, we won't reposition the MemoryStream back to the beginning, and thus subsequent Read operations will all return -1 because they're at the end of the MemoryStream. No?

Reply:
So what you're suggesting is that if CanSeek is false, after copying it to the new MemoryStream we always seek to the beginning of the stream? Or that when we copy the stream to the new MemoryStream we at least seek to the stream's original position if seekToOrigin is false? I think the second would be the right thing to do.

Reply:
That you set the MemoryStream's Position to 0 after the copy, regardless of seekToOrigin.  Otherwise, if seekToOrigin is false, the MemoryStream's Position will == MemoryStream.Length, and any attempts to Read from that will fail.

Of course, as I said, I don't know how this is used.  Maybe this issue never occurs.

Reply:
I don't think this issue occurs because of the way it is used, but I think is good to protect us from hitting that issue latter and spending time investigating why we're hitting such issue. Thanks for reviewing and the comment, will follow up in a PR.